### ISSUE TITLE
CAPTCHA Solving & Automation

### ISSUE THREAD

**Issue by JuroOravec** [JuroOravec/crawlee-one#34]
# CAPTCHA Solving & Automation

### 1. Used OCR approach for simple CAPTCHA

GuideChem uses a simple numeric CAPTCHA (e.g. "1497", "9537"). **Tesseract.js** with `tessedit_char_whitelist: '0123456789'` solves it reliably.

### Script: `scrapers/guidechem/scripts/ocr-captcha-test.ts`

- Same-session flow: load `/product/` first, then navigate to CAS page
- Screenshot `img.ure_lkh_yzm` → Tesseract OCR → fill `#delipcode` → submit
- Saves `scripts/captcha-debug.png` for inspection

```bash
pnpm exec tsx scripts/ocr-captcha-test.ts
# or: pnpm run ocr-captcha
```

**Result:** OCR solved "9537" correctly; form submission reached the CAS listing page.

---

### 2. Optical CAPTCHA solving toolbox (as per LLM)

For future targets with more advanced or distorted CAPTCHAs:

| Option | When to use | Tradeoffs |
|--------|-------------|-----------|
| **OCR (Tesseract)** | Simple numeric/text, clean fonts (like GuideChem) | Free, local, fast. Use `tessedit_char_whitelist: '0123456789'` for digits. |
| **Preprocessing + OCR** | OCR alone unreliable | Grayscale, binarize, denoise, deskew before Tesseract. |
| **CAPTCHA services** | Distorted fonts, reCAPTCHA, hCaptcha | 2Captcha, Anti-Captcha, CapMonster. ~$1–3/1000; 10–30s latency. |
| **Custom ML model** | Fixed style, high volume | Train CNN on screenshots. Fast, no per-request cost; needs labeled data. |
| **Fallback chain** | Best of both | Try OCR first; if confidence low or submit fails, call 2Captcha. |

---

## 3. Custom ML model — data, training, deployment

> just curiosity - when you say "custom ML model" - how many images would we probably need to train ML model like this?
>
> so I'm thinking, I recently saw a captcha that had 4-5 letter/numbers if varoius rotations and sizes. I could in theiory keep reloading the page to generate new images, and after say 500, create custom ones by resizing / distorting / rotating, right? In that case, how would you actually train the model and how long would it take? and what then *is* the model? a matrix that's loaded into some framework? or a binary? how would I *use* the model in actual code and feed the input into it and retrieve the output?

### How many images?

For a simple numeric CAPTCHA (4 digits, 0–9):

- **Minimum:** ~500–1,000 labeled images per digit with strong augmentation
- **Better:** ~2,000–5,000 per digit (8k–50k total) for decent accuracy
- **Production-grade:** ~10,000+ per class for 95%+ accuracy

**Rule of thumb:** 1,000–5,000 images per class. For 10 digits: **10k–50k labeled images**.

**Augmentation helps:** 500 base images + 10–20 variants each (rotate, resize, noise, blur) → ~5k–10k effective training set.

### Training

- **Framework:** PyTorch or TensorFlow (Keras), Python
- **Architecture:** Small CNN (e.g. ResNet-18) or sequence model (one output per character)
- **Training time:** Local GPU: minutes–hours; CPU: hours–day; Colab/cloud GPU: similar to local

### What is the model?

**Weights** (arrays of floats) saved to files:

- **PyTorch:** `.pt` / `.pth` — dict of tensors
- **TensorFlow:** `.h5` (Keras) or SavedModel directory
- **ONNX:** `.onnx` — portable, many runtimes support it

Typically tens of MB for a small CAPTCHA model.

### Using the model in Node.js / crawler

1. **ONNX Runtime (Node.js)** — Export PyTorch/TF → ONNX, run with `onnxruntime-node`. Pure Node, no Python at runtime.
2. **TensorFlow.js** — Convert to TF.js, run with `@tensorflow/tfjs-node`.
3. **Python subprocess** — Save image → call Python script → read stdout.
4. **HTTP API** — Small Python service loads model, exposes endpoint; crawler POSTs image, gets text.

**Example (Node + ONNX):**
```javascript
import * as ort from 'onnxruntime-node';
const session = await ort.InferenceSession.create('captcha_model.onnx');
const feeds = { input: new ort.Tensor('float32', imageData, [1, 1, 48, 120]) };
const results = await session.run(feeds);
const text = results.output.data.map(i => CHARS[i]).join('');
```

---

## 4. Projects that automate the pipeline

> Are there projects that automate this entire flow? that if a target website has built a custom simple character-based captcha, then a project where you just link the URL / image selector, and it does this entire flow automatically, so you can then only deploy an inference server for that one website and use it's API to resolve captchas spcific to the target website? (or something like that)

### What exists

**There is no "paste URL + image selector → get inference API" one-click tool.** Collection (reload page, screenshot, extract) is still manual. Training and deployment can be semi-automated.

### dddd_trainer (sml2h3) — ~1k stars

**What it does:** Train custom models for [ddddocr](https://github.com/sml2h3/ddddocr). Handles: create project → prepare data → cache → train (PyTorch/CNN/CRNN) → export ONNX. Integrates with ddddocr and `ocr_api_server` for deployment.

**Flow:**
1. `python app.py create my_captcha`
2. Provide images; labels from filenames (`abc12_hash.jpg`) or `labels.txt`
3. `python app.py cache my_captcha /path/to/images/`
4. `python app.py train my_captcha` — trains and exports ONNX
5. Use ONNX with ddddocr or `ocr_api_server` for inference

**Gap:** You still collect and label images yourself.

### hcaptcha-model-factory (CaptchaAgent)

- Full pipeline: PyTorch training, cluster-based auto labeling, ONNX export, inference server
- Oriented toward hCaptcha-style tasks; not generic "any custom captcha"

### CapMonster MCS (commercial)

- CapMonster2's "Module Creation Studio" — create recognition modules for unknown captchas
- Commercial, Windows-based

### ddddocr (sml2h3)

- Pre-trained solver with HTTP API. Self-host:
```bash
docker run -d -p 7777:7777 xzxxn777/ddddocr:latest
```
- Endpoints: `/classification` (OCR text), `/capcode`, etc.
- Works with custom models trained via **dddd_trainer**

### Practical path

**dddd_trainer + ddddocr** — closest to end-to-end:

1. Small collector script (Playwright: reload, screenshot, crop captcha, save as `{label}_hash.png`)
2. Fix labels as needed
3. Run dddd_trainer to train and export ONNX
4. Deploy via ddddocr's API server with custom model

Roughly one day of work to wire collection + training + deployment.
